View product

aerotecnia99 opened the pull request:
### Contexto
Crear vista para visualizar detalles de un producto, a la cual se puede acceder desde el index de productos clickeando el botón detalles de cada ítem.

### Qué se esta haciendo

- Se agregó un botón "Detalles" en el index de productos.
- Se creó un nuevo componente `item-details.vue` para la vista del producto.
- Se modificó el controlador de items `app/items_controller.rb` y `routes.rb` para enrutar el recurso.

#### En particular hay que revisar
La nueva vista de detalles del producto.

-----------
#### Info Adicional (pantallazos, links, fuentes, etc.)

gmq commented on app/javascript/components/item-detail.vue:
```diff
@@ -0,0 +1,55 @@
+        {{ item.description }}
+      </span>
+      <base-button
+        :variant="loading ? 'disabled' : 'primary'"
```
Un detalle es que el variant de `base-button` solo cambia los colores, para desactivarlo además tienes que agregar `disabled` como propiedad.

difernandez commented on app/javascript/components/item-list-card.vue:
```diff
@@ -38,18 +36,26 @@ async function buy() {
+      </button>
+      <a
+        :href="`/items/${itemId}`"
+        class="w-1/2 py-4 disabled:text-zinc-500"
```
mini detalle: el link nunca va a estar deshabilitado, entonces nunca se aplicaría `disabled:text-zinc-500`. Quizás se podría "deshabilitar" pasándo un `href` vacío si está el `loading`, y en ese mismo caso habilitar la clase `text-zinc-500` usando [binding de class](https://vuejs.org/guide/essentials/class-and-style.html#binding-html-classes)

difernandez commented on config/routes.rb:
```diff
@@ -14,6 +14,7 @@
 
   scope module: :app, defaults: { format: :html } do
     resources :purchases, only: [:index]
+    resources :items, only: [:index, :show]
```
el `index` no es estrictamente necesario, ya que ya está asociado al root más arriba

difernandez reviewed (commented):
Se ve bueno 💯 dejé un par de comentarios chicos para que veas

difernandez commented on app/javascript/components/base-button.vue:
```diff
@@ -9,19 +9,22 @@ const variants = {
 
 interface Props {
   variant?: keyof typeof variants
+  disabled?: boolean
```
Sobre lo que puso memo: en este caso no es necesario definir el disabled como prop ni ponerlo explícito en el `button` del template. En vue, los atributos que uno le ponen a una instancia de un componente se aplican sobre el primer elemento de su template, [más info acá](https://vuejs.org/guide/components/attrs.html#attribute-inheritance). Osea que basta con que en el `item-detail` le pases el disabled como lo tienes ahora no más.
Sería necesario definirlo como prop si vas a hacer algo más con el valor, usarlo de condición en un if por ejemplo, pero si es solo para pasarlo como atributo al primer elemento eso ya estaría cubierto

galvadiz replied:
ahhh ya, o sea que borraría todo lo relacionado a ```disabled``` de ```base-button``` y solo dejo lo de ```item-detail``` cierto?

difernandez replied:
claro, con eso debería quedar funcionando igual que ahora

difernandez reviewed (approved):
dejo un comentario chico, pero queda aprobado!

aerotecnia99 committed:
feature(items-index): details button added

aerotecnia99 committed:
feature(item-details): view details of specific product

style(item-details): deleting unused import

feature(item-details): view details of specific product

style(item-details): deleting unused import

monkeyci commented on app/javascript/application.js:
```diff
@@ -8,6 +8,8 @@ import BaseCheckbox from './components/base-checkbox.vue';
 import ItemList from './components/item-list.vue';
 import PurchaseList from './components/purchase-list.vue';
+import ItemDetails from './components/item-details.vue';
+import InlineSvg from 'vue-inline-svg';
```
🚫 **[eslint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Absolute imports should come before relative imports  import/imports-first

difernandez replied:
aquí hay que corregir lo que dice el mono 👮‍♂️ es mover este import para arriba, antes que los relativos. No recuerdo si está regla es auto-fixeable, pero si lo es, puedes [configurar vscode para que corrija lo que sea auto-fixeable al guardar el archivo](https://la-guia.platan.us/setup/configuracion_de_tu_entorno_local/herramientas/linters#linters:~:text=Puedes%20hacer%20que-,al%20guardar%20un%20archivo,-.js/%20.vue)

ldlsegovia reviewed (approved):
Se ve bien! quizás Diego te puede dar mejor feedback sobre esta parte que es pro en front.
Algo que sería deseable es que acomodes un poco el estilo para que funcione en otras resoluciones

![image](https://github.com/platanus/the-store1/assets/3026413/753b3a11-4957-4c62-9900-99991dbfc928)

Cuando hacés cosas de front está bueno acompañar la descripción del PR con capturas o mejor aún con gif. Yo suelo usar https://getkap.co/

difernandez commented on app/javascript/components/item-list-card.vue:
```diff
@@ -44,12 +48,20 @@ async function buy() {
+      </button>
+      <button
+        class="w-full py-4 text-blue-800 disabled:text-zinc-500"
+        @click="goToItemDetails"
```
si el `@click` de un botón solo redirige, en verdad debería ser un `<a>` con un href. Eso te permite cosas como abrir el link en otra pestaña, o copiarlo con el menú del botón derecho

difernandez commented on yarn.lock:
Me tinca que agregaste la librería con `npm install`, no? Por lo que veo, eso sale en la documentación. Pregunto porque casi todos los cambios en este archivo son cambios en el registro de que se saca, que pasó a ser `npmjs`. Nosotros usamos yarn, entonces cuando un paquete dice `npm install`, en verdad nosotro usamos `yarn add`. Prueba corriendo yarn install, creo que con eso se debería desaparecer estos cambios y volver a `yarnpkg`

difernandez commented on app/javascript/components/item-details.vue:
```diff
@@ -0,0 +1,71 @@
+        </p>
+        <button
+          class="w-full rounded-full bg-blue-800 py-2 px-4 text-lg font-bold text-white"
+          @click="buy"
```
acá agregaría un `:disabled="loading"` para tratar de evitar que se mande con doble click

difernandez commented on app/javascript/components/item-details.vue:
```diff
@@ -3,7 +3,6 @@ import { computed, ref } from 'vue';
 import { useNotification } from '@kyvg/vue3-notification';
 import type { Item } from '../api/items';
 import purchasesApi from '../api/purchases';
-import InlineSvg from 'vue-inline-svg';
```
tratamos de evitar agregar commits en un PR que cambien algo que se toca en un commit anterior del mismo PR. Está bien cuando el cambio es como agregar sobre lo que se hizo en otro commit, onda para ir construyendo una feature de a poco, pero cuando es un cambio que corrige un error del commit anterior, lo que hacemos es agregar ese cambio con un fixup al commit donde se introdujo ese cambio y luego agregarlo a ese commit con un rebase. Seguimos ese mismo procedimiento para aplicar cambios que nos piden en un PR, o si hay que corregir cosas del mono por ejemplo. Con esto queda la historia del PR más limpia y fácil de entender. Te recomendaría cambiarlo para que este cambio se aplique así, para que practiques lo que es los rebase también. [Acá hay un par de links útiles](https://la-guia.platan.us/setup/configuracion_de_proyectos/git#rebase), y si necesitas ayuda me dices tambn

difernandez reviewed (changes requested):
buena! algunos detalless

monkeyci commented on app/javascript/components/item-details.vue:
```diff
@@ -0,0 +1,71 @@
+<template>
+  <div>
+    <a
+      class="mb-12 inline-flex items-center rounded-full bg-gray-200 py-2 px-4 text-sm font-normal"
```
⚠️ **[eslint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Invalid Tailwind CSS classnames order  tailwindcss/classnames-order

monkeyci commented on app/javascript/components/item-details.vue:
```diff
@@ -0,0 +1,71 @@
+          {{ item.description }}
+        </p>
+        <button
+          class="w-full rounded-full bg-blue-800 py-2 px-4 text-lg font-bold text-white"
```
⚠️ **[eslint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Invalid Tailwind CSS classnames order  tailwindcss/classnames-order

monkeyci commented on app/javascript/components/item-list-card.vue:
```diff
@@ -44,12 +46,20 @@ async function buy() {
-    >
-      Comprar
-    </button>
+    <div class="inline-flex h-16 w-full items-start justify-around py-4 px-3">
```
⚠️ **[eslint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Invalid Tailwind CSS classnames order  tailwindcss/classnames-order

difernandez reviewed (approved):
dejo aprobado, pero échale un ojo a los últimos comentarios del mono! Esa es una regla que es muy latera de mantener sin el autofix de eslint
